libsupport/src/Plugin.cpp

```diff
@@ -0,0 +1,187 @@
+#include "katana/Plugin.h"
+
+#include <dlfcn.h>
+
+#include <cstdlib>
+#include <unordered_map>
+
+#include <boost/filesystem.hpp>
+#include <katana/Logging.h>
+#include <katana/Result.h>
+#include <katana/Strings.h>
+
+namespace {
+
+class PluginLoader {
+  std::unordered_map<std::string, katana::Plugin> plugins_;
+
+public:
+  void LoadPlugin(const boost::filesystem::path& path) {
+    if (plugins_.count(path.string())) {
+      return;
+    }
+
+    void* handle = dlopen(path.c_str(), RTLD_LAZY | RTLD_LOCAL);
+    if (!handle) {
+      KATANA_LOG_WARN(
+          "skipping plugin {}: failed to dynamically link plugin: {}", path,
+          katana::ResultErrno());
+      return;
+    }
+    const auto KatanaPluginInit =
+        reinterpret_cast<const katana::PluginMetadata* (*)()>(
+            dlsym(handle, "KatanaPluginInit"));
+    const katana::PluginMetadata* metadata = KatanaPluginInit();
+    if (!metadata) {
+      KATANA_LOG_WARN("plugin {} may have failed to load", path);
+    }
+    KATANA_LOG_VERBOSE("Loaded plugin: {}", path);
+    plugins_.emplace(
+        path.string(), katana::Plugin{
+                           .metadata = metadata,
+                           .so_path = path.string(),
+                           .so_handle = handle,
+                       });
+  }
+
+  void LoadAllPlugins() {
+    if (!plugins_.empty()) {
+      return;
+    }
+
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
+      for (const boost::filesystem::directory_entry& entry :
```

Comment:
The order is arbitrary (probably usually FS level dir entry order). I don't want to guarantee any plugin loading order yet. If we want to add that later we can. Doing it right would probably require dependency information or a way to directly control loading. Right now I'm intentionally being dumb and just loading everything that can be loaded. This defers decisions about what plugins are allowed to do.

Do you think this is a mistake?

Reply:
Doing dependencies seems like a lot of work. I was more wondering if we want the order to be deterministic or if we want to intentionally randomize the order to prevent people from depending on "natural" filesystem order.

Reply:
Yea. I would be temped to randomize to prevent people from making assumptions or naming plugins to cause a specific order (if we sorted by file name). I'll add randomization. Preventing assumptions now will be useful so we can add features later without needing to be backwards compatible with some early assumptions.